### PR TITLE
Feat/object store

### PR DESCRIPTION
**Pull request by deanls1**
please review my code

### CODE REVIEW

File: http/objectstore/iamapi/iam_api_server.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	apiRouter.Methods(http.MethodPost).Path("/change-password").HandlerFunc(iamApi.ChangePassword).Queries("newPassword", "{newPassword:.*}")
+	apiRouter.Methods(http.MethodPost).Path("/update-accessKey_status").HandlerFunc(iamApi.SetStatus).Queries("userName", "{userName:.*}", "status", "{status:.*}")
+
+	apiRouter.Methods(http.MethodPost).Path("/creat-group").HandlerFunc(iamApi.CreatGroup).Queries("groupName", "{groupName:.*}", "version", "{version:.*}")
+	apiRouter.Methods(http.MethodGet).Path("/get_group").HandlerFunc(iamApi.GetGroup).Queries("groupName", "{groupName:.*}", "version", "{version:.*}")
+	apiRouter.Methods(http.MethodPost).Path("/delete-group").HandlerFunc(iamApi.DeleteGroup).Queries("groupName", "{groupName:.*}", "version", "{version:.*}")
+	apiRouter.Methods(http.MethodGet).Path("/list-groups").HandlerFunc(iamApi.ListGroups).Queries("pathPrefix", "{pathPrefix:.*}")
+
+	apiRouter.NotFoundHandler = http.HandlerFunc(response.NotFoundHandler)
+}
```
Comment (qianhh) [inline http/objectstore/iamapi/iam_api_server.go:45]: iam这块api接口既然与s3接口是分离的，那么是否可以简单实现，只需要实现简单的几个必要接口，满足用户使用需求就好，并不完全按照aws的文档来做？

Comment (deanls1) [reply http/objectstore/iamapi/iam_api_server.go:45]: Okey, I will try to make the IAM API interface simple and meet user needs at the same time

File: http/objectstore/iam/check_handler_auth.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+		if err != nil {
+			log.Errorf("ReadAll err:%v", err)
+			return cred, owner, api_errors.ErrMalformedXML
+		}
+
+		// Populate payload to extract location constraint.
+		r.Body = ioutil.NopCloser(bytes.NewReader(payload))
+
+		// Populate payload again to handle it in HTTP handler.
+		r.Body = ioutil.NopCloser(bytes.NewReader(payload))
```
Comment (qianhh) [inline http/objectstore/iam/check_handler_auth.go:77]: Is this duplicate code?

Comment (deanls1) [reply http/objectstore/iam/check_handler_auth.go:77]: Yes,between these two lines, there are some functions before, I forgot to delete the corresponding code when I delete these functions,I will fix it in next submission.

File: http/objectstore/README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Object Store
+- Object Store 用户、地域、桶、对象等基本数据结构实现
+- Object Store 用户认证 api 实现
+- Object Store 桶相关操作 api 实现
+- Object Store 对象操作 api 实现
+- Object Store 权限操作 api 实现
```
Comment (qianhh) [inline http/objectstore/README.md:6]: Translate into English.

File: http/objectstore/s3api/s3_object_handler.go [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+		return
+	}
+}
+
+// HeadObjectHandler - HEAD Object
+//https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html
+// The HEAD operation retrieves metadata from an object without returning the object itself.
+func (s3a *s3ApiServer) HeadObjectHandler(w http.ResponseWriter, r *http.Request) {
+	bucket, object := getBucketAndObject(r)
+	var ctx = context.Background()
```
Comment (beeleelee) [inline http/objectstore/s3api/s3_object_handler.go:142]: Should avoid to use context.Background(), find a way to use context of the application daemon.

Comment (deanls1) [reply http/objectstore/s3api/s3_object_handler.go:142]: ”context.Background()“ is temporary and will be changed later

File: http/objectstore/iam/bucket-metadata-sys.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+func (sys *bucketMetadataSys) GetBucketMeta(bucket, username string, meta *BucketMetadata) error {
+	err := sys.db.Get(bucketPrefix+username+"-"+bucket, meta)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// HeadBucketMeta metadata for a bucket.
+func (sys *bucketMetadataSys) HeadBucketMeta(bucket, username string) bool {
```
Comment (beeleelee) [inline http/objectstore/iam/bucket-metadata-sys.go:125]: If use this method to tell whether a bucket has in your system, why name it HasBucketMeta?

Comment (deanls1) [reply http/objectstore/iam/bucket-metadata-sys.go:125]: It was used in "headbuckethandler" before, so it was named "headbucketmeta", now it is used in many functions, but I forgot to change the name, I will correct it later

File: cmd/objectstore/server.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	for _, ip := range utils.MustGetLocalIP4().ToSlice() {
+		log.Infof("start sever at http://%v%v", ip, os.Getenv(fileDagStoragePort))
+	}
+	err := http.ListenAndServe(os.Getenv(fileDagStoragePort), router)
+	if err != nil {
+		log.Errorf("Listen And Serve err%v", err)
+		return
+	}
+	defer uleveldb.DBClient.Close()
+}
```
Comment (qianhh) [inline cmd/objectstore/server.go:40]: How does OpenDb handle error cases? The defer statement should be written after DB has been successfully opened.

Comment (deanls1) [reply cmd/objectstore/server.go:40]: This was my oversight and will be fixed soon

File: http/objectstore/store/store.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		Writer:                     nil,
+		Reader:                     nil,
+		PutObjReader:               nil,
+		AccTime:                    time.Unix(0, 0).UTC(),
+		Legacy:                     false,
+		VersionPurgeStatusInternal: "",
+		VersionPurgeStatus:         "",
+		NumVersions:                0,
+		SuccessorModTime:           time.Now().UTC(),
+	}
```
Comment (beeleelee) [inline http/objectstore/store/store.go:60]: Need a reconsideration about how much the meta info fields we want to save to db. It seems too many.

File: http/objectstore/utils/string.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+//String return points
+func String(v string) *string {
+	return &v
+}
+
+//String return points
+func Time(v time.Time) *time.Time {
+	return &v
+}
```
Comment (qianhh) [inline http/objectstore/utils/string.go:13]: Are these two functions unnecessary and can be used directly with &?

**Review by qianhh — APPROVED** [review]

File: http/objectstore/store/store.go (hunk 7 above)
Comment (deanls1) [reply http/objectstore/store/store.go:60]: Okay, I'll check it out and re-determine what data we're going to store later

File: http/objectstore/utils/string.go (hunk 8 above)
Comment (deanls1) [reply http/objectstore/utils/string.go:13]: Yes, i will delete it later

**Pull request merged**